kv/dist_sender.go

```diff
@@ -555,32 +548,52 @@ func (ds *DistSender) Send(_ context.Context, call client.Call) {
 			var replicas replicaSlice
 			if err == nil {
 				replicas = newReplicaSlice(ds.gossip, desc)
-				rpcErr = ds.sendRPC(desc.RaftID, replicas, args, reply)
-				err = rpcErr
-			}
 
-			// If the RPC went through fine, an error (if any) is in the reply.
-			if err == nil {
-				err = reply.Header().GoError()
+				// Rearrange the replicas so that those replicas with long common
+				// prefix of attributes end up first. If there's no prefix, this is a
+				// no-op.
+				order := ds.optimizeReplicaOrder(replicas)
+
+				// Don't mess with replicas - we need the full list below for
+				// cache expiration considerations, but we may not want to send
+				// to all of them now.
+				useReplicas := replicas
+
+				// If this request needs to go to a leader and we know who that is, move
+				// it to the front and make it our single replica to send to.
+				// This allows us to special-case this when an RPC error comes back.
+				if !(proto.IsRead(args) && args.Header().ReadConsistency == proto.INCONSISTENT) &&
+					leader.StoreID > 0 {
+					if i := replicas.FindReplica(leader.StoreID); i >= 0 {
+						replicas.MoveToFront(i)
+						useReplicas = replicas[:1]
+						order = rpc.OrderStable
+					}
+				}
+
+				rpcErr = ds.sendRPC(desc.RaftID, useReplicas, order, args, reply)
 			}
 
 			// For an RPC error to occur, we've must've been unable to contact
-			// any replicas. In this case, likely all nodes are down; this
-			// should really only happen if either the replica set is dead or
+			// any replicas. In this case, either
+			// - we sent only to a leader or
+			// - all nodes are down.
+			// The second case only happens if either the replica set is dead or
 			// we have an old descriptor which does not have a list of current
 			// nodes; we hope it's the latter.
-			// TODO(tschottdorf): individual rpc calls always return an
-			// rpcError, which is always retriable. Thus, rpc.Send will always
-			// return retriable errors if everything is unreachable. That seems
-			// awkward; rpcError should probably not be retriable (rpc.Send
-			// will still succeed if it manages to talk to *anyone*), and
-			// then we can clear the cache here only for retriable errors.
-			// It amounts to the same situations though.
+			// In the first case, we update the leader cache only.
 			if rpcErr != nil {
-				ds.updateLeaderCache(proto.RaftID(desc.RaftID), proto.Replica{})
-				ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+				if len(replicas) == 1 && replicas[0].StoreID == leader.StoreID {
+					log.Warningf("RPC error to leader %d; expiring cache", leader.StoreID)
```

Comment:
It seems like this change up will result in worse performance. For one thing, an RPC error trying to talk to the designated leader will result in us continuing the retry loop in dist sender Send(). That will mean an exponential backoff.

Further, the rpc.Send() method is tuned to fairly quickly send of alternate PRCs (in SendNextTimeout) in the event that the first is taking too long. This change loses that functionality if there's a leader.